content/en/blog/_posts/2025-02-14-cloud-controller-manager-chicken-egg-problem.md

```diff
@@ -0,0 +1,124 @@
+
+For cloud-controller-managers running in the same cluster, they are managing.
+
+1. Enable host networking. In most cases, a cloud controller manager will need to communicate with an API service endpoint associated with the infrastructure. Setting “hostNetwork” to true will ensure that the cloud controller is using the host networking instead of the container network and, as such, will have the same network access as the host operating system. It will also remove the dependency on the networking plugin. This will ensure that the cloud controller has access to the infrastructure endpoint (always check your networking configuration against your infrastructure provider’s instructions).  
+2. Use a scalable resource type. `Deployments` and `DaemonSets` are useful for controlling the lifecycle of a cloud controller. They allow easy access to running multiple copies for redundancy as well as using the Kubernetes scheduling to ensure proper placement in the cluster. When using these primitives to control the lifecycle of your cloud controllers and running multiple replicas, you must remember to enable leader election, or else your controllers will collide with each other which could lead to nodes not being initialized in the cluster.  
+3. Target the controller manager containers to the control plane. There might exist other controllers which need to run outside the control plane (for example, Azure’s node manager controller). Still, the controller managers themselves should be deployed to the control plane. Use a node selector or affinity stanza to direct the scheduling of cloud controllers to the control plane to ensure that they are running in a protected space. Cloud controllers are vital to adding and removing nodes to a cluster as they form a link between Kubernetes and the physical infrastructure. Running them on the control plane will help to ensure that they run with a similar priority as other core cluster controllers and that they have some separation from non-privileged user workloads.  
+   1. It is worth noting that an anti-affinity stanza to prevent cloud controllers from running on the same host is also very useful to ensure that a single node failure will not degrade the cloud controller performance.  
+4. Ensure that the tolerations allow operation. Use tolerations on the manifest for the cloud controller container to ensure that it will schedule to the correct nodes and that it can run in situations where a node is initializing. This means that cloud controllers should tolerate the `node.cloudprovider.kubernetes.io/uninitialized` taint, and it should also tolerate any taints associated with the control plane (for example, `node-role.kubernetes.io/master` or `node-role.kubernetes.io/control-plane`). It can also be useful to tolerate the `node.kubernetes.io/not-ready` taint to ensure that the cloud controller can run even when the node is not yet available for health monitoring.
```

Comment:
- I'd really like to list `node-role.kubernetes.io/control-plane` ahead of the legacy role `node-role.kubernetes.io/master`.
- Officially, those are labels and not taints. You might be thinking of `node-role.kubernetes.io/control-plane:NoSchedule` and `node-role.kubernetes.io/master:NoSchedule`.

Reply:
good catch, thank you

Reply:
actually, for the not-ready, i was thinking about the taint. the one listed here https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/#taint-based-evictions